.travis.yml

```diff
@@ -1,33 +1,72 @@
 os: linux
 dist: focal
-language: c++
-env:
-  global:
-       - DEBIAN_FRONTEND: noninteractive
-addons:
-  apt:
-    packages:
-        - wget
-        - locales
-        - python3-pip
+language: minimal
+
+services:
+  - docker
 
-before_install:
-        - sudo locale-gen "en_US.UTF-8"
-        - export LANG="en_US.UTF-8"
-        - pip install git+https://github.com/antmicro/tuttest
-        - tuttest README.md unnamed0 | bash -
-jobs:
-  include:
-    - stage: SymbiFlow
-      install:
-            - tuttest README.md unnamed1 | bash -
+env:
+  # Testing QuickLogic toolchain on all OSes
+  - TOOLCHAIN=quicklogic      OS=ubuntu OS_VERSION=trusty
+  - TOOLCHAIN=quicklogic      OS=ubuntu OS_VERSION=xenial
+  - TOOLCHAIN=quicklogic      OS=ubuntu OS_VERSION=bionic
+  - TOOLCHAIN=quicklogic      OS=ubuntu OS_VERSION=eoan
+  - TOOLCHAIN=quicklogic      OS=ubuntu OS_VERSION=focal
+  - TOOLCHAIN=quicklogic      OS=centos OS_VERSION=6
+  - TOOLCHAIN=quicklogic      OS=centos OS_VERSION=7
+  - TOOLCHAIN=quicklogic      OS=centos OS_VERSION=8
+  # Testing Xilinx 7 Series toolchain on all OSes
+  - TOOLCHAIN=xilinx-series-7 OS=ubuntu OS_VERSION=trusty
+  - TOOLCHAIN=xilinx-series-7 OS=ubuntu OS_VERSION=xenial
+  - TOOLCHAIN=xilinx-series-7 OS=ubuntu OS_VERSION=bionic
+  - TOOLCHAIN=xilinx-series-7 OS=ubuntu OS_VERSION=eoan
+  - TOOLCHAIN=xilinx-series-7 OS=ubuntu OS_VERSION=focal
+  - TOOLCHAIN=xilinx-series-7 OS=centos OS_VERSION=6
+  - TOOLCHAIN=xilinx-series-7 OS=centos OS_VERSION=7
+  - TOOLCHAIN=xilinx-series-7 OS=centos OS_VERSION=8
 
-      script:
-            - tuttest README.md unnamed3 | bash -
-    - #
-      install:
-            - tuttest README.md unnamed2 | bash -
+install:
+ - if [ "$OS" = "ubuntu" ]; then export INSTALL_COMMAND="apt update && apt install -y"; fi
+ - if [ "$OS" = "centos" ]; then export INSTALL_COMMAND="yum -y install"; fi
+ - if [ "$OS" = "ubuntu" ]; then export INSTALL_PACKAGES="locales software-properties-common git wget xz-utils make build-essential zlib1g libz-dev libffi-dev ca-certificates"; fi
+ - if [ "$OS" = "centos" ]; then export INSTALL_PACKAGES="git wget make gcc-c++ zlib-devel xz libffi-devel perl-Pod-Html diffutils byacc"; fi
+ - export DOCKER_NAME=test
+ - export IN_DOCKER_EXEC="docker exec -it $DOCKER_NAME"
+ - docker create
+        --name $DOCKER_NAME
+        --env LANG="en_US.UTF-8"
+        --env LANGUAGE="en_US"
+        --env DEBIAN_FRONTEND="noninteractive"
+        -v /home/travis:/home/travis
+        --workdir $PWD
+        --interactive
+        --tty
+        $OS:$OS_VERSION
+ - docker start $DOCKER_NAME
+ - $IN_DOCKER_EXEC bash -c "$INSTALL_COMMAND $INSTALL_PACKAGES"
+ - if [ "$OS" = "ubuntu" ]; then $IN_DOCKER_EXEC locale-gen en_US.UTF-8; fi
+ - if [ "$OS" = "ubuntu" ]; then $IN_DOCKER_EXEC update-locale; fi
+ - $IN_DOCKER_EXEC wget https://www.openssl.org/source/openssl-1.1.1c.tar.gz
```

Comment:
WTF is going on here with openssl? There is something seriously wrong here?

Reply:
Tuttest is using dataclasses (https://docs.python.org/3/library/dataclasses.html) that were introduced in python 3.7. This version of python is not available on official ubuntu 14.04 repositories. It can be found on ``deadsnakes`` repository, but it is missing pip and SSL modules. PIP installation with python installed from the ``deadsnakes`` repo ends with the following error: 

```
docker exec -it test sudo python get-pip.py
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
...
Could not fetch URL https://pypi.org/simple/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
```

In order to build python with an ssl module, it requires an openssl: 1.0.2 or 1.1 compatible libssl with X509_VERIFY_PARAM_set1_host(), but Ubuntu 14.04 is shipped with OpenSSL version 1.0.1j. This forces me to rebuild openssl.

Reply:
We shouldn't be building Python, see my solution below.